Incorrectly warning about ignored files

After globs were added to the cli input in 5a599c70a98eaed0d43a8f28690921ef4ad0702d, my command to lint rule docs has broken:

```
DEBUG=eslint* node_modules/.bin/eslint --plugin markdown --no-eslintrc --config conf/rule-doc-config.json docs/rules/quote-props.md
  eslint:cli Running on files +0ms
  eslint:config Using command line config conf/rule-doc-config.json +3ms
  eslint:cli-engine Linting complete in: 3ms +10ms

/home/ian/Local_code/eslint/eslint/docs/rules/quote-props.md
  0:0  warning  File ignored because of your .eslintignore file. Use --no-ignore to override

✖ 1 problem (0 errors, 1 warning)
```

Compare to my output if I revert to the commit prior to globs being added:

```
▶ git reset --hard 37bb3159f748463861ba9aadf704e76af3c2537d
HEAD is now at 37bb315 Merge pull request #3518 from ev0rtex/update-expression-rule-logic

▶ DEBUG=eslint* node_modules/.bin/eslint --plugin markdown --no-eslintrc --config conf/rule-doc-config.json docs/rules/quote-props.md
  eslint:cli Running on files +0ms
  eslint:config Using command line config conf/rule-doc-config.json +3ms
  eslint:cli-engine Processing docs/rules/quote-props.md +10ms
  eslint:cli-engine Linting docs/rules/quote-props.md +1ms
  eslint:config Constructing config for /home/ian/Local_code/eslint/eslint/docs/rules/quote-props.md +0ms
  eslint:config Not using .eslintrc or package.json files +0ms
  eslint:config Merging command line config file +1ms
  eslint:config Merging command line environment settings +0ms
  eslint:config Merging command line plugins +0ms
  eslint:cli-engine Load plugin markdown +4ms

docs/rules/quote-props.md
    7:20  error  Unexpected token ;
  130:6   error  Unexpected string

✖ 2 problems (2 errors, 0 warnings)
```

Something appears to have changed in how files are determined to be ignored.  Furthermore, if I follow the warning's advice and add `--no-ignore`, the warning is not shown, but the file is still not linted:

```
▶ DEBUG=eslint* node_modules/.bin/eslint --plugin markdown --no-ignore --no-eslintrc --config conf/rule-doc-config.json docs/rules/quote-props.md
  eslint:cli Running on files +0ms
  eslint:config Using command line config conf/rule-doc-config.json +2ms
  eslint:cli-engine Linting complete in: 2ms +9ms
```

I'm running this from inside the eslint project.  I don't have ESLint installed globally, and with the `eslint-plugin-markdown` that I'm using, I've had to install eslint within eslint (which I do with an `npm link` and `npm link eslint`).  It's a little unorthodox, I know, but it was working just fine.  I could also have done:

```
DEBUG=eslint* node bin/eslint.js --plugin markdown --no-eslintrc --config conf/rule-doc-config.json docs/rules/quote-props.md            
  eslint:cli Running on files +0ms
  eslint:config Using command line config conf/rule-doc-config.json +3ms
  eslint:cli-engine Linting complete in: 3ms +11ms

/home/ian/Local_code/eslint/eslint/docs/rules/quote-props.md
  0:0  warning  File ignored because of your .eslintignore file. Use --no-ignore to override

✖ 1 problem (0 errors, 1 warning)
```

There is no eslintignore path being specified, so I'm not sure what you mean there.

--- a/lib/cli-engine.js
+++ b/lib/cli-engine.js
@@ -274,17 +274,37 @@ function isErrorMessage(message) {
  *
  * Also makes sure all path separators are POSIX style for `glob` compatibility.
  *
- * @param {string} pathname The directory path to be modified
- * @returns {string} The glob path or the file path itself
+ * @param {string[]} [extensions] An array of accepted extensions
+ * @returns {Function} A function that takes a pathname and returns a glob that
+ *                     matches all files with the provided extensions if
+ *                     pathname is a directory.
  */
-function processPath(pathname) {
-    var newPath = pathname;
+function processPath(extensions) {
+    var suffix = "/**";
 
-    if (shell.test("-d", pathname)) {
-        newPath = path.join(pathname, "**");
+    if (extensions) {
+        if (extensions.length === 1) {
+            suffix += "/*." + extensions[0];
+        } else {
+            suffix += "/*.{" + extensions.join(",") + "}";
+        }
     }
 
-    return newPath.replace(/\\/g, "/");
+    /**
+     * A function that converts a directory name to a glob pattern
+     *
+     * @param {string} pathname The directory path to be modified
+     * @returns {string} The glob path or the file path itself
+     */
+    return function(pathname) {
+        var newPath = pathname;
+
+        if (shell.test("-d", pathname)) {
+            newPath = pathname + suffix;
+        }
+
+        return newPath.replace(/\\/g, "/");
+    };
 }
 
 
@@ -409,18 +429,19 @@ CLIEngine.prototype = {
 
     /**
      * Executes the current configuration on an array of file and directory names.
-     * @param {string[]} files An array of file and directory names.
+     * @param {string[]} patterns An array of file and directory names.
      * @returns {Object} The results for all files that were linted.
      */
-    executeOnFiles: function(files) {
+    executeOnFiles: function(patterns) {
         var results = [],
             processed = {},
             options = this.options,
             fileCache = this._fileCache, // eslint-disable-line no-underscore-dangle
             configHelper = new Config(options),
-            ignoredPaths = options.ignore ? IgnoredPaths.load(options).patterns : [],
+            ignoredPaths = IgnoredPaths.load(options),
+            ignoredPathsList = ignoredPaths.patterns || [],
             extensions = options.extensions.map(function(ext) {
-                return ext.charAt(0) === "." ? ext : "." + ext;
+                return ext.charAt(0) === "." ? ext.substr(1) : ext;
             }),
             globOptions,
             stats,
@@ -428,13 +449,13 @@ CLIEngine.prototype = {
 
         startTime = Date.now();
 
-        files = files.map(processPath);
-        ignoredPaths = ignoredPaths.map(processPath);
+        patterns = patterns.map(processPath(extensions));
+        ignoredPathsList = ignoredPathsList.map(processPath());
 
         globOptions = {
             nodir: true,
             realpath: true,
-            ignore: ignoredPaths
+            ignore: ignoredPathsList
         };
 
         /**
@@ -444,7 +465,7 @@ CLIEngine.prototype = {
          * @returns {void}
          */
         function executeOnFile(filename) {
-            if (extensions.indexOf(path.extname(filename)) === -1 || processed[filename]) {
+            if (processed[filename]) {
                 return;
             }
 
@@ -496,15 +517,19 @@ CLIEngine.prototype = {
             results.push(res);
         }
 
-        files.forEach(function(pattern) {
-            glob.sync(pattern, globOptions).forEach(executeOnFile);
+        patterns.forEach(function(pattern) {
+            if (shell.test("-f", pattern) && !ignoredPaths.contains(pattern)) {
+                executeOnFile(fs.realpathSync(pattern));
+            } else {
+                glob.sync(pattern, globOptions).forEach(executeOnFile);
+            }
         });
 
         // only warn for files explicitly passed on the command line
         if (options.ignore) {
-            files.forEach(function(file) {
+            patterns.forEach(function(file) {
                 var fullPath = path.resolve(file);
-                if (shell.test("-e", fullPath) && !processed[fullPath]) {
+                if (shell.test("-f", fullPath) && !processed[fullPath]) {
                     results.push(createIgnoreResult(file));
                 }
             });

--- a/tests/lib/cli-engine.js
+++ b/tests/lib/cli-engine.js
@@ -304,6 +304,15 @@ describe("CLIEngine", function() {
             assert.equal(report.results[1].warningCount, 0);
         });
 
+        it("should process when file is given by not specifying extensions", function() {
+
+            engine = new CLIEngine();
+
+            var report = engine.executeOnFiles(["tests/fixtures/files/foo.js2"]);
+            assert.equal(report.results.length, 1);
+            assert.equal(report.results[0].messages.length, 0);
+        });
+
         it("should return zero messages when given a config with environment set to browser", function() {
 
             engine = new CLIEngine({

